src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -2306,6 +2311,886 @@ internal unsafe void CM(B b)
+    static void Caller(delegate*<string, void> ptr) => ptr(""1"");
+    static void Main()
+    {
+        Caller(&M);
+    }
+}", expectedOutput: "1");
+
+            verifier.VerifyIL("C.Main()", expectedIL: @"
+{
+  // Code size       12 (0xc)
+  .maxstack  1
+  IL_0000:  ldftn      ""void C.M(string)""
+  IL_0006:  call       ""void C.Caller(delegate*<string,void>)""
+  IL_000b:  ret
+}
+");
+        }
+
+        [Fact]
+        public void AddressOf_CannotAssignToVoidStar()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    static void M()
+    {
+        void* ptr = &M;
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (6,21): error CS0428: Cannot convert method group 'M' to non-delegate type 'void*'. Did you intend to invoke the method?
+                //         void* ptr = &M;
+                Diagnostic(ErrorCode.ERR_MethGrpToNonDel, "&M").WithArguments("M", "void*").WithLocation(6, 21)
+            );
+        }
+
+        [Fact]
+        public void AddressOf_DisallowedInExpressionTree()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+using System;
+using System.Linq.Expressions;
+unsafe class C
+{
+    static string M1(delegate*<string> ptr) => ptr();
+    static string M2() => string.Empty;
+
+    static void M()
+    {
+        Expression<Func<string>> a = () => M1(&M2);
+        Expression<Func<string>> b = () => (&M2)();
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (11,48): error CS8775: '&' on method groups cannot be used in expression trees
+                //         Expression<Func<string>> a = () => M1(&M2);
+                Diagnostic(ErrorCode.ERR_AddressOfMethodGroupInExpressionTree, "M2").WithLocation(11, 48),
+                // (12,44): error CS0149: Method name expected
+                //         Expression<Func<string>> b = () => (&M2)();
+                Diagnostic(ErrorCode.ERR_MethodNameExpected, "(&M2)").WithLocation(12, 44)
+            );
+        }
+
+        [Fact]
+        public void AmbiguousApplicableMethodsAreFilteredForStatic()
+        {
+            var verifier = CompileAndVerifyFunctionPointers(@"
+using System;
+interface I1{}
+interface I2
+{
+    string Prop { get; }
+}
+
+public unsafe class C : I1, I2 {
+    void M(I1 i) {}
+    static void M(I2 i) => Console.Write(i.Prop);
+    public static void Main() {
+        delegate*<C, void> a = &M;
+        a(new C());
+    }
+    public string Prop { get => ""I2""; }
+}", expectedOutput: "I2");
+
+            verifier.VerifyIL("C.Main()", expectedIL: @"
+{
+  // Code size       19 (0x13)
+  .maxstack  2
+  .locals init (delegate*<C,void> V_0)
+  IL_0000:  ldftn      ""void C.M(I2)""
+  IL_0006:  stloc.0
+  IL_0007:  newobj     ""C..ctor()""
+  IL_000c:  ldloc.0
+  IL_000d:  calli      0x4
+  IL_0012:  ret
+}
+");
+        }
+
+        [Fact]
+        public void UnconstrainedGenericNotAllowed()
```

Comment:
>Unconstrained [](start = 20, length = 13)

It doesn't look like constraints have anything to do with the errors for this scenario.

Reply:
Constraints was the wrong word, will rename these.